tests/MethodCallTest.php

```diff
@@ -264,15 +264,14 @@ public static function main(): void {
             ],
             'dateTimeImmutableStatic' => [
                 '<?php
-                    /** @psalm-immutable */
                     final class MyDate extends DateTimeImmutable {}
 
                     $today = new MyDate();
                     $yesterday = $today->sub(new DateInterval("P1D"));
 
                     $b = (new DateTimeImmutable())->modify("+3 hours");',
                 'assertions' => [
-                    '$yesterday' => 'MyDate|false',
+                    '$yesterday' => 'MyDate',
```

Comment:
Unsure about this change: is `MyDate|false` actually expected? From a type perspective, PHP 8 doesn't seem to have `false` as a possible return value anymore :thinking:

Reply:
`getTimestamp` was fixed by #7855, I was intending to go fix the others but I never got around to it. I think we might need to fix that in the CallMap for functions not defined in the stub, and for the stubbed functions we'll probably have to use the `|false` return and copy/paste it to Php80.phpstub without the `|false` return.

Reply:
The wording above confused me more :D 

If I understand it correctly:

 * the stub represents current (latest) PHP state
 * the `CallMap` includes PHP-version-specific overrides

If it's the opposite (stubs represent "oldest" state, CallMap represents "newest" state), then it's super-confusing overall :D

Reply:
The latter, CallMap represents newest state, CallMap_historical represents oldest state, and there are diffs for each PHP version to get from CallMap to CallMap_historical. For stubs, stubs represent oldest state except for PHPxx.phpstub, which is applied after the other stubs to add changes for newer versions. It's confusing.

Reply:
> stubs represent oldest state

Argh @_@

Reply:
I think the reason it was done that way is because we don't actually have a way to "unapply" a stub, but for the callmap it's easy. It's unfortunate that they're opposite though...

Reply:
stubs override Callmap though, so if you have a function declared in stubs, it doesn't really matter what you have in the Callmap (or Callmap_historical)

Reply:
Oh, yeah, sorry if that wasn't clear.

Reply:
I'm a bit lost on what to do - anyone able to pick this up perhaps? :| 

If the stub takes precedence, is this patch acceptable?

Reply:
Sorry for the confusion, here's a recap:
- stubs override callmap so your stub here is good
- you added new methods in the existing stub, so it overrided existing definition in callmap (that's why you saw types change in tests)
- there is a specific stub file for each minor version of php (it is mainly used for new classes and stuff in new version)
- this means your stub here applies to any php version, which is not great because types changed between versions
- however, we do have meta variables made to be used in cases like these: https://github.com/vimeo/psalm/blob/e0acf22e402d45d1aebb75f4365ebf701fabce2f/stubs/CoreGenericFunctions.phpstub#L1096

So, could you please adapt your stub to use conditional returns in order to reflect the correct return given a php version and it should be good :)

Reply:
Oh, I didn't even know about using `PHP_MAJOR_VERSION` in a conditional return, that's neat. I assumed it would have to be overridden in PHP80.phpstub.

Reply:
Yeah, we also have this weird one too: https://github.com/vimeo/psalm/blob/b4fdc3e326fea865de49827f723a2703bb18617a/src/Psalm/Internal/PhpVisitor/Reflector/FunctionLikeDocblockScanner.php#L504

But useful in some edge cases :)

Reply:
With your example though, won't that break when PHP 9.0 is released? Wouldn't you want to invert it and do `(PHP_MAJOR_VERSION is 7 ...)`? (assuming we don't care about 5.x at all anymore).

Reply:
Some users, including some maintainers won't be happy if you break 5.x analysis :p. But yeah, we're bound to have some work when PHP 9 will come anyway. If it can be handled now it's better, otherwise we'll do it later

Reply:
:o we can use this too: https://github.com/vimeo/psalm/blob/b49682a4853803f72996fbd6664b5e69a75472ae/tests/Template/ConditionalReturnTypeTest.php#L866

I didn't even know it worked with int ranges! That's better than the previous solution even!

Reply:
Oh, cool, that's a much better solution! I was thinking that wasn't possible because you have to use `is` and you can't say `PHP_MAJOR_VERSION < 8`.

Reply:
Ok, I've documented all types:

 * looked at core PHP implementation
 * looked at core PHP stubs
 * implemented conditional types based on major PHP version (luckily, only one case of that)

Still one problem left, but this thread is resolved :-)

Reply:
Note: each added commit tries to reference the sources of my findings.

Reads: avoid squashing, please :P